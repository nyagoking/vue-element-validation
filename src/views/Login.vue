<template>
  <el-card class="login-form">
    <div>LOGIN</div>
    <div>
      <el-form ref="loginForm" :rules="rules" :model="loginModel">
        <el-form-item label="ID" prop="id">
          <el-input v-model="loginModel.id" class="login-id"></el-input>
        </el-form-item>
        <el-form-item label="PASSWORD" prop="password">
          <el-input v-model="loginModel.password" class="login-password" show-password></el-input>
        </el-form-item>
        <el-row class="button-area">
          <el-button class="login-btn" type="primary" :disabled="!validationResult" @click="login">
            Login
          </el-button>
        </el-row>
      </el-form>
    </div>
  </el-card>
</template>

<script lang="ts">
import ElementUi from 'element-ui';
import { ElForm, ValidateCallback } from 'element-ui/types/form';
import Vue from 'vue';
import Component from 'vue-class-component';

interface LoginModel {
  id: string;
  password: string;
}

Vue.use(ElementUi);

@Component
export default class Login extends Vue {
  /* data */
  public $refs!: { [key: string]: ElForm };
  private loginModel: LoginModel | null = null;
  private validationResult: boolean = false;
  private data() {
    return {
      rules: {
        id: [{ required: true, message: 'required', trigger: 'blur' }],
        password: [{ required: true, message: 'required', trigger: 'blur' }],
      },
    };
  }

  /* lifecycle hooks */
  private created() {
    this.loginModel = {
      id: '',
      password: '',
    };
  }
  private beforeUpdate() {
    this.updateValidationResult();
  }

  /* methods */
  private login() {
    alert('Login clicked!');
  }
  private updateValidationResult() {
    const callback: ValidateCallback = (isValid: boolean, invalidFields: object) => {
      this.validationResult = isValid;
    };

    this.$refs.loginForm.validate(callback);
  }
}
</script>

<style lang="scss">
.login-form {
  width: 480px;
  margin: auto;
}
</style>
