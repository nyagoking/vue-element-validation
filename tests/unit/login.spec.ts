import { createLocalVue, mount, shallowMount, Wrapper, BaseWrapper } from '@vue/test-utils';
import ElementUI from 'element-ui';
import { ElForm, ValidateCallback } from 'element-ui/types/form';
import Login from '@/views/Login.vue';
import VueRouter from 'vue-router';

const localVue = createLocalVue();
localVue.use(ElementUI);
localVue.use(VueRouter);

describe('Login.vue', () => {
  const router = new VueRouter();

  let wrapper: Wrapper<Login>;

  beforeEach(() => {
    wrapper = shallowMount(Login);
  });

  afterEach(() => {
    expect(wrapper.element).toMatchSnapshot();
    jest.clearAllMocks();
  });

  describe('初期化', () => {
    describe('template', () => {
      beforeEach(() => {
        wrapper = mount(Login, {
          localVue,
          router,
        });
      });

      describe('入力フィールド(ID)', () => {
        it('値が空であること', () => {
          const id = wrapper.find('.login-id input').element as HTMLInputElement;

          expect(id.value).toBe('');
        });
      });

      describe('入力フィールド(PASSWORD)', () => {
        it('値が空であること', () => {
          const password = wrapper.find('.login-password input').element as HTMLInputElement;

          expect(password.value).toBe('');
        });
      });

      describe('ログインボタン', () => {
        it('不活性であること', () => {
          const button: BaseWrapper = wrapper.find('.login-btn');

          expect(button.props('disabled')).toBe(true);
        });
      });
    });

    describe('$data', () => {
      it('プロパティ loginModel にデフォルト値を設定すること', () => {
        const expected = {
          id: '',
          password: '',
        };

        expect(wrapper.vm.$data.loginModel).toEqual(expected);
      });

      it('プロパティ validationResult に false を設定すること', () => {
        expect(wrapper.vm.$data.validationResult).toBe(false);
      });
    });
  });

  describe('lifecycle hooks', () => {
    let form: ElForm;
    let handlers: (() => void)[] | undefined;
    let mockValidate: jest.Mock<any>;

    describe('beforeUpdate', () => {
      beforeEach(() => {
        form = wrapper.vm.$refs.loginForm as ElForm;
        handlers = wrapper.vm.$options.beforeUpdate as (() => void)[] | undefined;
        mockValidate = jest.fn();
        form.validate = mockValidate;
      });

      function execBeforeUpdate() {
        if (handlers && handlers.length === 1) {
          handlers[0].call(wrapper.vm);
        } else {
          fail('lifecycle hook "beforeUpdate" should be defined.');
        }
      }

      it('コールバックを引数にフォームのバリデーションを実施すること', () => {
        execBeforeUpdate();
        expect(mockValidate).toHaveBeenCalledWith(expect.any(Function));
      });

      describe('callback', () => {
        let callback: ValidateCallback;

        beforeEach(() => {
          execBeforeUpdate();
          callback = mockValidate.mock.calls[0][0];
        });

        it('第1引数 isValid がバリデーション正常(true)ならプロパティ validationResult に true を設定すること', () => {
          wrapper.setData({ validationResult: false });

          callback(true, {});
          expect(wrapper.vm.$data.validationResult).toBe(true);
        });

        it('第1引数 isValid がバリデーションエラー(false)ならプロパティ validationResult に false を設定すること', () => {
          wrapper.setData({ validationResult: true });

          callback(false, { id: [{ message: 'required', field: 'id' }] });
          expect(wrapper.vm.$data.validationResult).toBe(false);
        });
      });
    });
  });
});
