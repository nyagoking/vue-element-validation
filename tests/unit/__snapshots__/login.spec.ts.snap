// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue lifecycle hooks beforeUpdate callback 第1引数 isValid がバリデーションエラー(false)ならプロパティ validationResult に false を設定すること 1`] = `
<el-card-stub
  class="login-form"
>
  <div>
    LOGIN
  </div>
   
  <div>
    <el-form-stub
      labelsuffix=""
      model="[object Object]"
      rules="[object Object]"
      showmessage="true"
      validateonrulechange="true"
    >
      <el-form-item-stub
        inlinemessage=""
        label="ID"
        prop="id"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-id"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-form-item-stub
        inlinemessage=""
        label="PASSWORD"
        prop="password"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-password"
          showpassword="true"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-row-stub
        align="top"
        class="button-area"
        justify="start"
        tag="div"
      >
        <el-button-stub
          class="login-btn"
          disabled="true"
          icon=""
          nativetype="button"
          type="primary"
        >
          
          Login
        
        </el-button-stub>
      </el-row-stub>
    </el-form-stub>
  </div>
</el-card-stub>
`;

exports[`Login.vue lifecycle hooks beforeUpdate callback 第1引数 isValid がバリデーション正常(true)ならプロパティ validationResult に true を設定すること 1`] = `
<el-card-stub
  class="login-form"
>
  <div>
    LOGIN
  </div>
   
  <div>
    <el-form-stub
      labelsuffix=""
      model="[object Object]"
      rules="[object Object]"
      showmessage="true"
      validateonrulechange="true"
    >
      <el-form-item-stub
        inlinemessage=""
        label="ID"
        prop="id"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-id"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-form-item-stub
        inlinemessage=""
        label="PASSWORD"
        prop="password"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-password"
          showpassword="true"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-row-stub
        align="top"
        class="button-area"
        justify="start"
        tag="div"
      >
        <el-button-stub
          class="login-btn"
          icon=""
          nativetype="button"
          type="primary"
        >
          
          Login
        
        </el-button-stub>
      </el-row-stub>
    </el-form-stub>
  </div>
</el-card-stub>
`;

exports[`Login.vue lifecycle hooks beforeUpdate コールバックを引数にフォームのバリデーションを実施すること 1`] = `
<el-card-stub
  class="login-form"
>
  <div>
    LOGIN
  </div>
   
  <div>
    <el-form-stub
      labelsuffix=""
      model="[object Object]"
      rules="[object Object]"
      showmessage="true"
      validateonrulechange="true"
    >
      <el-form-item-stub
        inlinemessage=""
        label="ID"
        prop="id"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-id"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-form-item-stub
        inlinemessage=""
        label="PASSWORD"
        prop="password"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-password"
          showpassword="true"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-row-stub
        align="top"
        class="button-area"
        justify="start"
        tag="div"
      >
        <el-button-stub
          class="login-btn"
          disabled="true"
          icon=""
          nativetype="button"
          type="primary"
        >
          
          Login
        
        </el-button-stub>
      </el-row-stub>
    </el-form-stub>
  </div>
</el-card-stub>
`;

exports[`Login.vue 初期化 $data プロパティ loginModel にデフォルト値を設定すること 1`] = `
<el-card-stub
  class="login-form"
>
  <div>
    LOGIN
  </div>
   
  <div>
    <el-form-stub
      labelsuffix=""
      model="[object Object]"
      rules="[object Object]"
      showmessage="true"
      validateonrulechange="true"
    >
      <el-form-item-stub
        inlinemessage=""
        label="ID"
        prop="id"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-id"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-form-item-stub
        inlinemessage=""
        label="PASSWORD"
        prop="password"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-password"
          showpassword="true"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-row-stub
        align="top"
        class="button-area"
        justify="start"
        tag="div"
      >
        <el-button-stub
          class="login-btn"
          disabled="true"
          icon=""
          nativetype="button"
          type="primary"
        >
          
          Login
        
        </el-button-stub>
      </el-row-stub>
    </el-form-stub>
  </div>
</el-card-stub>
`;

exports[`Login.vue 初期化 $data プロパティ validationResult に false を設定すること 1`] = `
<el-card-stub
  class="login-form"
>
  <div>
    LOGIN
  </div>
   
  <div>
    <el-form-stub
      labelsuffix=""
      model="[object Object]"
      rules="[object Object]"
      showmessage="true"
      validateonrulechange="true"
    >
      <el-form-item-stub
        inlinemessage=""
        label="ID"
        prop="id"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-id"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-form-item-stub
        inlinemessage=""
        label="PASSWORD"
        prop="password"
        showmessage="true"
      >
        <el-input-stub
          autocomplete="off"
          class="login-password"
          showpassword="true"
          type="text"
          validateevent="true"
          value=""
        />
      </el-form-item-stub>
       
      <el-row-stub
        align="top"
        class="button-area"
        justify="start"
        tag="div"
      >
        <el-button-stub
          class="login-btn"
          disabled="true"
          icon=""
          nativetype="button"
          type="primary"
        >
          
          Login
        
        </el-button-stub>
      </el-row-stub>
    </el-form-stub>
  </div>
</el-card-stub>
`;

exports[`Login.vue 初期化 template ログインボタン 不活性であること 1`] = `
<div
  class="el-card login-form is-always-shadow"
>
  <!---->
  <div
    class="el-card__body"
  >
    <div>
      LOGIN
    </div>
     
    <div>
      <form
        class="el-form"
      >
        <div
          class="el-form-item is-required"
        >
          <label
            class="el-form-item__label"
            for="id"
          >
            ID
          </label>
          <div
            class="el-form-item__content"
          >
            <div
              class="login-id el-input"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="text"
              />
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
         
        <div
          class="el-form-item is-required"
        >
          <label
            class="el-form-item__label"
            for="password"
          >
            PASSWORD
          </label>
          <div
            class="el-form-item__content"
          >
            <div
              class="login-password el-input el-input--suffix"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="password"
              />
              <!---->
              <span
                class="el-input__suffix"
              >
                <span
                  class="el-input__suffix-inner"
                >
                  <!---->
                  <!---->
                  <!---->
                  <!---->
                </span>
                <!---->
              </span>
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
         
        <div
          class="button-area el-row"
        >
          <button
            class="el-button login-btn el-button--primary is-disabled"
            disabled="disabled"
            type="button"
          >
            <!---->
            <!---->
            <span>
              
          Login
        
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Login.vue 初期化 template 入力フィールド(ID) 値が空であること 1`] = `
<div
  class="el-card login-form is-always-shadow"
>
  <!---->
  <div
    class="el-card__body"
  >
    <div>
      LOGIN
    </div>
     
    <div>
      <form
        class="el-form"
      >
        <div
          class="el-form-item is-required"
        >
          <label
            class="el-form-item__label"
            for="id"
          >
            ID
          </label>
          <div
            class="el-form-item__content"
          >
            <div
              class="login-id el-input"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="text"
              />
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
         
        <div
          class="el-form-item is-required"
        >
          <label
            class="el-form-item__label"
            for="password"
          >
            PASSWORD
          </label>
          <div
            class="el-form-item__content"
          >
            <div
              class="login-password el-input el-input--suffix"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="password"
              />
              <!---->
              <span
                class="el-input__suffix"
              >
                <span
                  class="el-input__suffix-inner"
                >
                  <!---->
                  <!---->
                  <!---->
                  <!---->
                </span>
                <!---->
              </span>
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
         
        <div
          class="button-area el-row"
        >
          <button
            class="el-button login-btn el-button--primary is-disabled"
            disabled="disabled"
            type="button"
          >
            <!---->
            <!---->
            <span>
              
          Login
        
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Login.vue 初期化 template 入力フィールド(PASSWORD) 値が空であること 1`] = `
<div
  class="el-card login-form is-always-shadow"
>
  <!---->
  <div
    class="el-card__body"
  >
    <div>
      LOGIN
    </div>
     
    <div>
      <form
        class="el-form"
      >
        <div
          class="el-form-item is-required"
        >
          <label
            class="el-form-item__label"
            for="id"
          >
            ID
          </label>
          <div
            class="el-form-item__content"
          >
            <div
              class="login-id el-input"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="text"
              />
              <!---->
              <!---->
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
         
        <div
          class="el-form-item is-required"
        >
          <label
            class="el-form-item__label"
            for="password"
          >
            PASSWORD
          </label>
          <div
            class="el-form-item__content"
          >
            <div
              class="login-password el-input el-input--suffix"
            >
              <!---->
              <input
                autocomplete="off"
                class="el-input__inner"
                type="password"
              />
              <!---->
              <span
                class="el-input__suffix"
              >
                <span
                  class="el-input__suffix-inner"
                >
                  <!---->
                  <!---->
                  <!---->
                  <!---->
                </span>
                <!---->
              </span>
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
         
        <div
          class="button-area el-row"
        >
          <button
            class="el-button login-btn el-button--primary is-disabled"
            disabled="disabled"
            type="button"
          >
            <!---->
            <!---->
            <span>
              
          Login
        
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
